File: test/model/wardrobe-paginator-test.js
var assert = buster.referee.assert;
var refute = buster.referee.refute;
var WardrobePaginator = DUB.model.WardrobePaginator;

function createData(size) {
	var data = [];
	for (var i = 1; i <= size; i++) {
		data.push(i);
	}
	return data;
}

buster.testCase("Wardrobe Paginator tests", {
	"paginator returns same page if at end of data": function () {
		var paginator = new WardrobePaginator(createData(16));
		var page1 = paginator.current();
		var page2 = paginator.next();
		var page3 = paginator.next();
		assert.isArray(page1);
		refute.equals(page1, page2);
		assert.equals(page2, page3);
	},
	
	"paginator has 15 as default page size": function () {
		var paginator15 = new WardrobePaginator(createData(15));
		var paginator16 = new WardrobePaginator(createData(16));
		
		assert.equals(paginator15.current().length, 15);
		assert.equals(paginator16.current().length, 15);
	},
	
	"previous returns same page if at beginning of data": function() {
		var paginator = new WardrobePaginator(createData(16));
		paginator.next();
		var page = paginator.previous();
		assert.equals(paginator.previous(), page);
	},
	
	"current returns current page": function () {
		var paginator = new WardrobePaginator(createData(33));
		assert.equals(paginator.current(), [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
		
		var page = paginator.next();
		assert.equals(paginator.current(), page);
		
		var page = paginator.next();
		assert.equals(paginator.current(), page);
	},
	
	"next returns next set of data": function () {
		var paginator = new WardrobePaginator(createData(10), 3);
		assert.equals(paginator.current(), [1, 2, 3]);
		assert.equals(paginator.next(), [4, 5, 6]);
		assert.equals(paginator.next(), [7, 8, 9]);
		assert.equals(paginator.next(), [10]);
	},
	
	"previous returns previous set of data": function () {
		var paginator = new WardrobePaginator(createData(10), 3);
		paginator.next();
		paginator.next();
		assert.equals(paginator.next(), [10]);
		assert.equals(paginator.previous(), [7, 8, 9]);
		assert.equals(paginator.previous(), [4, 5, 6]);
		assert.equals(paginator.previous(), [1, 2, 3]);
	}
});

File: Assets/test/Editor/WardrobePaginatorTest.cs
using NUnit.Framework;

[TestFixture]
public class WardrobePaginatorTest {

	private int[] createData(int size) {
		int[] data = new int[size];
		for (int i = 1; i <= size; i++) {
			data[i - 1] = i;
		}
		return data;
	}

	[Test]
	public void PaginatorReturnsSamePageIfAtEndOfData() {
		WardrobePaginator<int> paginator = new WardrobePaginator<int>(createData(16));
		int[] page1 = paginator.Current();
		int[] page2 = paginator.Next();
		int[] page3 = paginator.Next();
		Assert.AreNotEqual(page1, page2);
		Assert.AreEqual(page2, page3);
	}

	[Test]
	public void PaginatorHas15AsDefaultPageSize() {
		WardrobePaginator<int> paginator15 = new WardrobePaginator<int>(createData(15));
		WardrobePaginator<int> paginator16 = new WardrobePaginator<int>(createData(16));

		Assert.AreEqual(15, paginator15.Current().Length);
		Assert.AreEqual(15, paginator16.Current().Length);
	}

	[Test]
	public void PreviousReturnsSamePageIfAtBeginningOfData() {
		WardrobePaginator<int> paginator = new WardrobePaginator<int>(createData(16));
		paginator.Next();
		int[] page = paginator.Previous();
		Assert.AreEqual(page, paginator.Previous());
	}

	[Test]
	public void CurrentReturnsCurrentPage() {
		WardrobePaginator<int> paginator = new WardrobePaginator<int>(createData(33));
		Assert.AreEqual(new int[] {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}, paginator.Current());

		int[] page = paginator.Next();
		Assert.AreEqual(page, paginator.Current());

		page = paginator.Next();
		Assert.AreEqual(page, paginator.Current());
	}

	[Test]
	public void NextReturnsNextSetOfData() {
		WardrobePaginator<int> paginator = new WardrobePaginator<int>(createData(10), 3);
		Assert.AreEqual(new int [] {1, 2, 3}, paginator.Current());
		Assert.AreEqual(new int [] {4, 5, 6}, paginator.Next());
		Assert.AreEqual(new int [] {7, 8, 9}, paginator.Next());
		Assert.AreEqual(new int[] { 10 }, paginator.Next());
	}

	[Test]
	public void PreviousReturnsPreviousSetOfData() {
		WardrobePaginator<int> paginator = new WardrobePaginator<int>(createData(10), 3);
		paginator.Next();
		paginator.Next();
		Assert.AreEqual(new int [] {10}, paginator.Next());
		Assert.AreEqual(new int [] {7, 8, 9}, paginator.Previous());
		Assert.AreEqual(new int [] {4, 5, 6}, paginator.Previous());
		Assert.AreEqual(new int[] { 1, 2, 3 }, paginator.Previous());
	}
}
